### PR TITLE
Extend StreamExt with state type parameter

### PR DESCRIPTION
**Pull request by djc**
@sfackler what do you think of something like this? I wanted this so I can use `futures::stream::iter()` but for a `StateStream`; this seemed like a neater solution than copying `iter()`.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -369,21 +369,21 @@ pub fn unfold<T, F, Fut, It, St>(init: T, f: F) -> Unfold<T, F, Fut>
...
     type Error = S::Error;
 
     #[inline]
-    fn poll(&mut self) -> Poll<StreamEvent<S::Item, ()>, S::Error> {
+    fn poll(&mut self) -> Poll<StreamEvent<S::Item, T>, S::Error> {
         self.0.poll().map(|a| {
             match a {
                 Async::Ready(Some(i)) => Async::Ready(StreamEvent::Next(i)),
-                Async::Ready(None) => Async::Ready(StreamEvent::Done(())),
+                Async::Ready(None) => Async::Ready(StreamEvent::Done(self.1.take().unwrap())),
```
Comment (sfackler) [inline src/lib.rs]: Could you replace the `unwrap` with an `expect("poll called after completion")` or something like that?

**Comment by sfackler** [comment]
Sounds reasonable to me!

**Comment by djc** [comment]
Good call, amended to include that change. (I seem to have trouble keeping my rustfmt install running, so it might not be formatted according to your standards -- sorry.)

**Comment by sfackler** [comment]
Thanks!

**Pull request merged**